webpack/assets/javascripts/react_app/common/hooks/API/APIHooks.js

```diff
@@ -36,7 +36,7 @@ export const useAPI = (method, url, options) => {
   }, [dispatch, url, method, options]);
 
   const response = useSelector(state =>
-    selectAPIResponse(state, keyRef.current)
+    selectAPIResponse(state, keyRef.current, options?.asMutable)
```

Comment:
can you show a use-case of the hook with asMutable?
for some reason, when testing it with [AuditCard](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/HostDetails/Audits/index.js#L43) which currently is the only usage of the hook,
I got an endless loop on the host details page,
passing `useAPI('get', url, { asMutable: true });`

Reply:
this isn't related to this PR, but for useAPI's useEffect referential equalities, the options object is a dependency, and in your example on every render a new object has been created, which causes the infinite loop.

```js
const API_OPTIONS = { asMutable: true } // created only once
const Component = () => { 
  useAPI('get', '/some-url', API_OPTIONS )
}
```

if the options depend on some props, we can memorize the obj reference with `useMemo` 
```js

const Component = ({ prop }) => { 
  const options = useMemo(() => ({key:  prop}), [prop]) // the reference keep the same 
  useAPI('get', '/some-url', API_OPTIONS )
}
```

Reply:
I think it is a bit of extra work for the consumer, maybe it can get resolved if we don't pass the options as a dependency in the hook's use-effect? we can destruct it and then pass specific keys into the API action?

Reply:
I agree with @laviro, it is unintuitive to require the options array to be a constant or memoized value by the consumer rather than being passed as an inline object.

Reply:
This issue is a general concern
#### TL;DR :)
We can make it simpler for consumption but with a tradeoff - the effect will be much less dynamic 

This is how the useAPI's `useEffect` looks:
```js
  useEffect(() => {
    if (url && method) {
      dispatch(
        APIActions[method]({
          url,
          ...options,
          key: keyRef.current,
        })
      );
    }
  }, [dispatch, url, method, options]);
```

these are the optional values under `options`:
```js
options = {
      key,
      url,
      headers = {},
      params = {},
      actionTypes = {},
      handleError = noop,
      handleSuccess = noop,
      successToast,
      errorToast,
      payload = {},
      updateData,
      asMutable // new
}
```

So even extracting each field one by one still creates an infinite loop because most of these fields  are not primitives (so `===` comparison still gives `false`)

Suppressing the options object from the dependency array breaks the` exhaustive-deps`, it's basically an  [anti-pattern](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) and might cause odd issues.

Using `useMemo` for avoiding useEffect's infinite loop caused by referential equalities is also been suggested by [Dan Abramov](https://overreacted.io/a-complete-guide-to-useeffect/), so when it is used moderately and it needed to keep the component dynamic, it is a common practice. 

But, for making it easier for consumers  there are some alternatives that I can think of:
####  extract the action dipsatch from the effect
```js
  function fetchAPI() {};
  useEffect(() => {
    if (url && method && key) {
      fetchAPI()
      );
    }
  }, [url, method, key]);
```
This approach comes with a  disadvantage - `options` have to be constant and can't be dynamic and derived by props,  the `useEffect` won't be triggered after a change.
####  using useRef
```js
  const optionsRef = useRef(option)
  useEffect(() => {
    if (url && method) {
      dispatch(
        APIActions[method]({
          url,
          ...options.current,
          key: keyRef.current,
        })
      );
    }
  }, [dispatch, url, method, options.current]);
```
This makes the `options` object to be a constant, but we get the same disadvantage from the previous one.

Do we want to keep options dynamic? IMHO it makes sense that `params`  or handlers  (i.e success/error) be dynamic and derived by other props, after all, there is a cost for generalizing API calls.  I agree that we can extract some fields from it, like the  `key` (and  `asMutable`), this would be a 

```js
useAPI(method, url, key, APIOptions, storeOptions) 
// APIOptions - options that are part of the API action
// storeOptions - other options  that aren't related to the API action. like `asMutable`
```

Reply:
@amirfefer could we perhaps use `useMemo` inside the hook for the options so that only when they really change we trigger a new call? and perhaps split those that aren't related to the api call like `asMutable` inside the hook?

Reply:
For most cases it's won't be possible -  `useMemo` uses a dependency array, which depend on the consumer's props and aren't passed to the `useAPI` hook:
```js
const ConsumerComponent = ({ a, b}) => {
  // params depend on props `a` and `b` which aren't passed to useAPI hook
  useAPI('get', '/url', { params: {...a, ...b}})
}
```

```js
const ConsumerComponent = ({ a, b}) => {
  // APIOptions are memorized by `a` and `b` props
  const APIOptions = useMemo(() => ({ parans: {...a, ...b} }), [a, b])
  useAPI('get', '/url', APIOptions)
}
```

Reply:
@amirfefer @tbrisker  what if we just remove the options object from the hook's useeffect deps?
I guess that it's only eslint that won't be satisfied.. though that would stop the endless loop.

I rhink the question is when do we want to trigger another render / API call, and that usually happens when one of the API request params has changed, right?

I would suggest to add as a third argument to the hook, the `params` object,  and as the fourth one the options argument.

IMO options shouldn't be part of the hooks effect deps, and for the params changes - we could Json.stringify them and pass the string as a deps. In params you have only key value pairs so that ehould be OK in my opinion.

Reply:
> I guess that it's only eslint that won't be satisfied.. 

It's not safe, omitting items from the deps list might end with stale closures,  especially when`useAPI` is a general hook that we can't control its use cases.

> I think the question is when do we want to trigger another render / API call, and that usually happens when one of the API request params has changed, right?

yes, but there are more edge cases for avodiding reduanant API call, for example, when some handlers might be affected by the conusmer's props:
```js
const ConsumerComponent = ({ successString }) => {
   const successCallback = () => successString
  // handleSuccess depend on prop successString
  useAPI('get', '/url', { handleSuccess: successCallback})
}
```
This example needs memorization, so the API won't be triggered redundantly multiple times (or even an infinite loop)

Reply:
@laviro , @tbrisker 
Following this discussion, I've created this https://github.com/theforeman/foreman/pull/8713, a suggestion for optimizing `useAPI`